src/typechecker.c

```diff
@@ -890,7 +889,35 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
     } break;
 
     /// Make sure a cast is even possible.
-    case NODE_CAST: TODO();
+    case NODE_CAST: {
+      Type *t_to = expr->type;
+      if (type_is_incomplete(t_to))
+        ERR(t_to->source_location, "Can not cast to incomplete type %T", t_to);
+
+      if (!typecheck_expression(ast, expr->cast.value))
+        return false;
+
+      Type *t_from = expr->cast.value->type;
+
+      // TODO: What do we do when from and to types are equal?
+
+      if (type_is_incomplete(t_from))
+        ERR(expr->cast.value->source_location, "Can not cast from an incomplete type %T", t_from);
+
+      // FROM any integer type TO any integer type is ALLOWED
+      if (is_integer(t_from) && is_integer(t_to)) break;
+
+      // FROM any pointer type TO any integer type is ALLOWED
+      if (is_pointer(t_from) && is_integer(t_to)) break;
+
+      // FROM any integer type TO any pointer type is DISALLOWED
+      if (is_integer(t_from) && is_pointer(t_to))
+        ERR(expr->source_location, "Can not cast from integer type %T to pointer type %T", t_from, t_to);
+
+      // TODO: arrays, functions, function pointers...
```

Comment:
I’d say:
- Casting between arrays is nonsense, just how you can’t e.g. cast a struct to another struct in C. However, casting between pointers to arrays would ofc be allowed.
- A function pointer is just another pointer.
- Casting to a function could be useful for performing some kind of ‘manual overload resolution’. Howeer, I think it only makes sense if the expression being cast is an overload set. If you want to cast something else to a random function type, you should have to cast it to a function pointer instead imo.

Reply:
Also, from any pointer to any pointer should also be allowed I’d say

Reply:
Well, I'm reluctant because there is kind of an interesting thing going on right now in that the only way to generate a pointer is by taking the address of something . Which means that, inherently, at least at the time the pointer is created and until it goes out of scope, what it's pointing to is valid, and dereferencing it won't cause memory issues. It would be really cool if we could retain this sort of "feature" as best as possible, even though I'm really bad at explaining it haha. I do realize that array subscripting already breaks this feature; however, I do think it wouldn't be too hard in sema to ensure array accesses are within bounds when compile-time known. It would be really cool, as well, to have the ability to add in checks into the runtime itself that calls some handler when an out of bounds access happens on a given array 🤔 

Sorry for rant, feel free to ignore, lmao. You are most likely right we should just allow explicit pointer to pointer casting if the base type is the same size or smaller, right?

Reply:
Only allowing valid pointers is kind of a massive headache (escape analysis especially) and nigh-on impossible once external libraries come into play... even C++ references, which are much safer than pointers in that regard, can be dangling references, and I’ve actually had a bug because of that before that was caused by, you’ve guessed it, a third-party library.

The only language that is pretty close to being able to do something like that is Rust, and idk about you but I don’t feel like implementing a fully-fledged borrow checker. And also, even Rust cannot make do w/o `unsafe {}` blocks...

So in conclusion, it’s a nice idea, but implementing it would be so much of a headache that allowing reinterpret casts wouldn’t be that big of a deal imo. And not having reinterpret casts *can* get rather annoying when you do end up needing them...

Reply:
I agree that array/slice accesses should be bounds-checked by default tho.

Reply:
> I do think it wouldn't be too hard in sema to ensure array accesses are within bounds when compile-time known.

If the subscript is a constant expression then sure, but often it won’t be, so in the general case, we’d have to do runtime checks...

Reply:
It's not a hard rule (like in Rust), but I would prefer that the default way of writing code makes it very hard to shoot yourself in the foot

Reply:
> make it genuinely hard to get a program with such obvious memory errors as dereferencing a compile time known zero address or something

Idk about ‘genuinely hard’, but I suppose we can add some checks. As for reinterpret casts, I feel like we’re gonna need them at some point, but we’ll see...

Reply:
Reinterpret casts I'm all good with:
> we should just allow explicit pointer to pointer casting if the base type is the same size or smaller, right?

Reply:
In that case we’d also have to check for alignment, don’t forget about that one :eyes:

Reply:
A larger type will always align with a smaller type, right? Like `2` aligns into `4`, etc

Reply:
Not necessarily. Alignment is not tied to size. You could have a 1 byte type w/ 16 byte alignment... for whatever reason. It’s not likely but it’s *possible*